packages/menu/src/Menu/Menu.tsx

```diff
@@ -336,50 +161,61 @@ export const Menu = React.forwardRef<HTMLDivElement, MenuProps>(function Menu(
       : { spacing, usePortal }),
   };
 
-  const providerData = useMemo(() => {
-    return { theme, darkMode };
-  }, [theme, darkMode]);
-
   const popoverContent = (
-    <MenuContext.Provider value={providerData}>
-      <Popover
-        key="popover"
-        active={open}
-        align={align}
-        justify={justify}
-        refEl={refEl}
-        adjustOnMutation={adjustOnMutation}
-        {...popoverProps}
+    <DescendantsProvider
+      context={MenuDescendantsContext}
+      descendants={descendants}
+      dispatch={dispatch}
+    >
+      <MenuContext.Provider
+        value={{
+          theme,
+          darkMode,
+          highlightIndex,
+        }}
       >
-        <div
-          className={cx(
-            rootMenuStyle,
-            rootMenuThemeStyles[theme],
-            css`
-              max-height: ${maxMenuHeightValue};
-            `,
-            className,
-          )}
-          ref={forwardRef}
+        <Popover
+          key="popover"
+          active={open}
+          align={align}
+          justify={justify}
+          refEl={refEl}
+          adjustOnMutation={adjustOnMutation}
+          onEntered={handlePopoverOpen}
+          {...popoverProps}
         >
-          {/* Need to stop propagation, otherwise Menu will closed automatically when clicked */}
-          {/* eslint-disable-next-line jsx-a11y/click-events-have-key-events*/}
-          <ul
-            {...rest}
-            className={scrollContainerStyle}
-            role="menu"
-            onClick={e => e.stopPropagation()}
-            ref={popoverRef}
+          <div
+            className={cx(
+              rootMenuStyle,
+              rootMenuThemeStyles[theme],
+              css`
+                max-height: ${maxMenuHeightValue};
+              `,
+              className,
+            )}
+            ref={forwardRef}
           >
-            {updatedChildren}
-          </ul>
-        </div>
-      </Popover>
-    </MenuContext.Provider>
+            {/* Need to stop propagation, otherwise Menu will closed automatically when clicked */}
+            {/* eslint-disable-next-line jsx-a11y/click-events-have-key-events*/}
+            <ul
+              {...rest}
+              className={scrollContainerStyle}
+              role="menu"
+              onClick={e => e.stopPropagation()}
```

Comment:
what is the reason we don't also want the same `stopPropagation` for keyboard events?

Reply:
This is here to prevent `useBackdropClick` from firing when this click propagates up. We don't need to handle other keyboard events since `useBackdropClick` only handles click & mousedown